huaweicloud/services/iotda/resource_huaweicloud_iotda_device.go

```diff
@@ -454,15 +484,43 @@ func buildDeviceCreateParams(d *schema.ResourceData) *model.AddDeviceRequest {
 			AuthInfo:    buildAuthInfo(d.Get("secret").(string), d.Get("fingerprint").(string), d.Get("secure_access").(bool)),
 		},
 	}
+	if extensionInfo != nil {
+		req.Body.ExtensionInfo = &extensionInfo
+	}
+	return &req
+}
+
+func buildDeviceShadowParams(d *schema.ResourceData, desiredRaw []interface{}) *model.UpdateDeviceShadowDesiredDataRequest {
+	rst := make([]model.UpdateDesired, len(desiredRaw))
+	for i, v := range desiredRaw {
+		desireds := v.(map[string]interface{})
+		desiredInfo := desireds["desired"]
+		rst[i] = model.UpdateDesired{
+			ServiceId: desireds["service_id"].(string),
+		}
+		if desiredInfo != nil {
+			rst[i].Desired = &desiredInfo
+		}
+	}
+
+	req := model.UpdateDeviceShadowDesiredDataRequest{
+		DeviceId: d.Id(),
+		Body: &model.UpdateDesireds{
+			Shadow: &rst,
+		},
+	}
+
 	return &req
 }
 
 func buildDeviceUpdateParams(d *schema.ResourceData) *model.UpdateDeviceRequest {
+	extensionInfo := utils.ValueIgnoreEmpty(d.Get("extension_info"))
 	req := model.UpdateDeviceRequest{
 		DeviceId: d.Id(),
 		Body: &model.UpdateDevice{
-			DeviceName:  utils.StringIgnoreEmpty(d.Get("name").(string)),
-			Description: utils.StringIgnoreEmpty(d.Get("description").(string)),
+			DeviceName:    utils.StringIgnoreEmpty(d.Get("name").(string)),
+			Description:   utils.StringIgnoreEmpty(d.Get("description").(string)),
+			ExtensionInfo: &extensionInfo,
```

Comment:
The return value of method `utils.ValueIgnoreEmpty()` may be nil, and this line of code has the risk of null pointer.